### PR TITLE
[RLlib]: Doubly Robust Off-Policy Evaluation

### PR DESCRIPTION
**Pull request by ghost**
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/rollout_worker.py
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:723]: Actually, I think these shorts are not necessary anymore. We only ever supported "is" and "wis" and now that we have the new API, we should just force users to specify a class here.

File: rllib/evaluation/rollout_worker.py
Comment (sven1977) [inline rllib/evaluation/rollout_worker.py:752]: Nice catch! I guess multi-agent offline RL would be very very tricky :)

File: rllib/offline/estimators/test_ope.py
Comment (sven1977) [inline rllib/offline/estimators/test_ope.py:30]: Let's use the new TrainerConfig objects from here on.

E.g.:
```
from ray.rllib.algorithms.dqn import DQNConfig

config = DQNConfig().offline_data(input_=...m ).framework(framework_str="torch").rollouts(batch_mode="...").exploration(exploration_config={})
trainer = config.build(env="CartPole-v0")
```

**Review by sven1977 — COMMENTED** [review]
Awesome PR @rapotdar ! Really impressive work.

A couple of top level thoughts (don't have to do this in this PR, but these should be follow up PRs):
* Overhaul of (existing?) documentation on OPEs. E.g. let's add a section on how to write your own.
* Warning if OPEs used inside training worker set with advice on how to best set these up inside the eval worker set.
* Unify config handling analogous to how we already configure e.g. exploration and replay buffers (dict with "type" + c'tor kwargs): e.g.:
```
config:
  off_policy_estimators:
    my_is:
      type: ImportanceSampling
      ctor_kwarg1: [some value1]
    dr:
      type: DoublyRobust
      ctor_kwarg1: [some value]
      ctor_kwarg2: [some other value]
```

Then this would produce a result dict with the OPE results under the user-given sub-keys ("my_is" and "dr").

^ just an example of how we could do this.

* how can we write tests that confirm that all the OPEs are working? Evaluate their outputs against a real env again? :)

File: rllib/evaluation/rollout_worker.py
Comment (rapotdar) [reply rllib/evaluation/rollout_worker.py:723]: They're necessary for the command-line API e.g. ```rllib train```, which is currently used in the [Offline RL docs](https://docs.ray.io/en/latest/rllib/rllib-offline.html), but I can definitely remove them instead.

File: rllib/evaluation/rollout_worker.py
Comment (rapotdar) [reply rllib/evaluation/rollout_worker.py:752]: Yup, this check was already in the OPE class, we can probably allow multi-agent in a future PR.

**Comment by rapotdar** [comment]
> * how can we write tests that confirm that all the OPEs are working? Evaluate their outputs against a real env again? :)
On this, we can add `off_policy_evaluation_method = ["simulation"]`. This logs `episode_reward_mean` from an actual simulated env, but we can compare that with the discounted estimates returned by OPE.

File: rllib/evaluation/rollout_worker.py
Comment (gjoliver) [reply rllib/evaluation/rollout_worker.py:723]: should we create a simple tune.registry for this? that would look pretty nice.
python object in config sometimes has downside, like I hope we don't have any python obj values in our policy_config ...

File: rllib/offline/estimators/direct_method.py
Comment (gjoliver) [inline rllib/offline/estimators/direct_method.py:78]: as we discussed offline, can we add a TODO here saying that potentially, this can be handled by AIR cross-validation?
Then the Direct estimator here can be a simple 1-time training run.

File: rllib/offline/estimators/direct_method.py
Comment (gjoliver) [inline rllib/offline/estimators/direct_method.py:91]: maybe we should follow python style guide about the variable names, not like following what's in the paper.

File: rllib/offline/estimators/doubly_robust.py
Comment (gjoliver) [inline rllib/offline/estimators/doubly_robust.py:14]: remove?

File: rllib/offline/estimators/direct_method.py
Comment (gjoliver) [inline rllib/offline/estimators/direct_method.py:54]: remove the Doubly Robust paper here?

File: rllib/offline/estimators/doubly_robust.py
Comment (gjoliver) [inline rllib/offline/estimators/doubly_robust.py:21]: same here about the k_fold_cv feature.

File: rllib/offline/estimators/doubly_robust.py
Comment (gjoliver) [inline rllib/offline/estimators/doubly_robust.py:27]: someone who is not super familiar with SampleBatch would think train_episodes[0] here is duplicated twice in the concatenated batch.
but actually concat_samples() is a static method ...
can you simply use SampleBatch.concat_samples() instead?

File: rllib/offline/estimators/doubly_robust.py
Comment (gjoliver) [inline rllib/offline/estimators/doubly_robust.py:50]: can you do
```
for t in reversed(range(episode.count)):
```
instead?

File: rllib/offline/estimators/qreg_tf_model.py
Comment (gjoliver) [inline rllib/offline/estimators/qreg_tf_model.py:14]: actually does this need to be a ModelV2?

File: rllib/offline/estimators/qreg_torch_model.py
Comment (gjoliver) [inline rllib/offline/estimators/qreg_torch_model.py:65]: I don't know if it's too much to ask. what do you think if we add some unit tests for this model? especially the train() impl?

**Review by gjoliver — COMMENTED** [review]
super cool man!

File: rllib/offline/estimators/direct_method.py
Comment (rapotdar) [reply rllib/offline/estimators/direct_method.py]: Fixed, thanks :)

File: rllib/offline/estimators/qreg_tf_model.py
Comment (rapotdar) [reply rllib/offline/estimators/qreg_tf_model.py]: No, I had made them ModelV2 for consistency earlier but have since removed it

File: rllib/offline/estimators/direct_method.py
Comment (ArturNiederfahrenhorst) [inline rllib/offline/estimators/direct_method.py:58]: Some args missing here

File: rllib/offline/estimators/direct_method.py
Comment (ArturNiederfahrenhorst) [inline rllib/offline/estimators/direct_method.py:63]: Although I like this, common you'd find prepended stars, as in `**kwargs`, in other places. So maybe we should keep this uniform.

File: rllib/algorithms/cql/tests/test_cql.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/cql/tests/test_cql.py:54]: Nit: I feel like making users name all estimators serves a niche case where people actually want to use the same estimator multiple times. 

I'd like the following better and still allow an optional _name argument.
[ 
   ("is", {}}, 
   (CustomOPE, {"arg1": 123}), 
   (CustomOPE, {"arg1": 456, "_name": "the_second_one"}
]

Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/cql/tests/test_cql.py:54]: I'd also like to get rid of users specifying estimators by string.
I know that this look convenient for us, but it's just less explicit and leaves room for error.
We should imho, where ever we can comfortably do this, make users interact with our classes and function directly rather then describing them by strings.

File: doc/source/rllib/rllib-training.rst
Comment (sven1977) [inline doc/source/rllib/rllib-training.rst:577]: Should this be better:

```
"off_policy_estimation_methods": {
  "is": {
    "type": ImportanceSampling,
    [room for possible c'top kwargs here],
  },
  "wis": {
    "type": WeightedImportanceSampling,
    [room for possible c'top kwargs here],
  }
}
```
?

Then in the results, the user would find the data under the self-defined keys "is" and "wis" (the user can change these).

File: rllib/BUILD
Comment (sven1977) [inline rllib/BUILD:1119]: Here, this seems already in the correct format (see my comment above).

File: rllib/agents/trainer_config.py
Comment (sven1977) [inline rllib/agents/trainer_config.py:910]: I'm actually not sure about this option right now. Do we still need it?
If we say that these estimator methods should - preferrably - be run on the evaluation worker track, and the user wanted to just see the live-env evaluation (actual rewards), then we wouldn't need estimators, but just do config.input_ = "sampler" on the evaluation workers, no?

File: rllib/agents/trainer_config.py
Comment (sven1977) [inline rllib/agents/trainer_config.py:912]: This is great! Can you add that `ope_method_name` may be any user-defined string?

File: rllib/offline/estimators/direct_method.py
Comment (ArturNiederfahrenhorst) [reply rllib/offline/estimators/direct_method.py:85]: (missing ticks)

File: rllib/offline/estimators/doubly_robust.py
Comment (ArturNiederfahrenhorst) [inline rllib/offline/estimators/doubly_robust.py:29]: # noqa: F841 not needed here I think?

File: rllib/algorithms/marwil/marwil.py
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:106]: Good.

File: rllib/algorithms/cql/tests/test_cql.py
Comment (sven1977) [reply rllib/algorithms/cql/tests/test_cql.py:54]: Not sure, actually. The user would have to know that there is this hidden `_name` arg (which is not a c'tor kwarg of the given class!).
I do prefer the clearer solution discussed above:

```
off_policy_estimation_methods={
  "user-provided-name-1": {
    "type": [some class or class-str],
    [c'tor kwargs],
  },
  "user-provided-name-2": {
    "type": [some class or class-str],
    [c'tor kwargs],
  },
}
```

This way, it fully matches our replay buffer configs as well as our exploration configs.

**Review by ArturNiederfahrenhorst — APPROVED** [review]
The only "serious" comment I have is about the test.
Love your work Rohan!

File: rllib/offline/estimators/direct_method.py
Comment (sven1977) [inline rllib/offline/estimators/direct_method.py:101]: Add an error here explaining that this estimator type is not available if `config.framework_str != "torch"`.

File: rllib/offline/estimators/test_ope.py
Comment (sven1977) [inline rllib/offline/estimators/test_ope.py:1]: Nice, thanks for adding unit tests to this entire complex!

File: rllib/offline/estimators/test_ope.py
Comment (sven1977) [inline rllib/offline/estimators/test_ope.py:64]: Nice, I like the way this looks now.

File: rllib/offline/estimators/qreg_torch_model.py
Comment (sven1977) [reply rllib/offline/estimators/qreg_torch_model.py:93]: @rapotdar and I were discussing offline to add "learning" tests for (at least the new) methods. Something like:
- Learn CartPole with some offline data for some time.
- Estimate the learnt policy.
- Check, whether estimation is correct by comparing it with the actual performance of the learnt policy in an actual env.

One big problem with this is variance. But I think it would be really realy cool to have these tests, even if the fault tolerance has to be large. Would also allow us to compare different OPE methods with each other.

Comment (sven1977) [reply rllib/offline/estimators/qreg_torch_model.py:93]: We can do this in a follow up PR, I think to not make this one blow up. :)

File: rllib/offline/estimators/fqe_torch_model.py
Comment (sven1977) [inline rllib/offline/estimators/fqe_torch_model.py:18]: Could you move this to the __init__'s docstring?

```Initializes a FQETorchModel instance.

Args:
   ...
"""
```

**Review by sven1977 — COMMENTED** [review]
Looks really great @rapotdar ! Almost ready to merge, just a few nits, comments, questions to address.

File: rllib/agents/trainer_config.py
Comment (ArturNiederfahrenhorst) [reply rllib/agents/trainer_config.py:913]: That's also part of what I was talking about in the resolved comment further up.
We have the same "issue" at multiple point in our library and I think we should, as you said, allow only full class paths.

But ontop of that provide a helper method that converts config objects between something "yamleable" and objects with actual classes in them. Also because I've ran into issues with providing the StorageUnit enumeration in our config. We can use this method ourselves inside config objects to go back and forth prior to execution time.

File: rllib/offline/estimators/test_ope.py
Comment (sven1977) [reply rllib/offline/estimators/test_ope.py:1]: Actually, can we:
* move this into a new sub-dir `esimators/tests`
* add this to our BUILD file right above:

```
<- HERE
py_test(
    name = "offline/evaluators/tests/test_ope",
    tags = ["team:ml", "offline"],
    size = "small",
    srcs = ["offline/evaluators/tests/test_ope.py"]
)
<- END: HERE

# --------------------------------------------------------------------
# Policies
# rllib/policy/
#
# Tag: policy
# --------------------------------------------------------------------

py_test(
    name = "policy/tests/test_compute_log_likelihoods",
    tags = ["team:ml", "policy"],
    size = "medium",
    srcs = ["policy/tests/test_compute_log_likelihoods.py"]
)

```

File: rllib/agents/trainer_config.py
Comment (rapotdar) [reply rllib/agents/trainer_config.py:940]: I'd say let's keep it for now, and I'll fix it in the next PR when I actually move the OPE to the eval workers?

File: rllib/offline/estimators/direct_method.py
Comment (rapotdar) [reply rllib/offline/estimators/direct_method.py:101]: Is this different from the `assert framework==`torch``? Technically, I think you can actually run this with policy as any PolicyV2.

File: rllib/agents/trainer_config.py
Comment (rapotdar) [reply rllib/agents/trainer_config.py:945]: The problem is I don't think the full classpath works with the `rllib train` utility. I can look at fixing it using an `importlib.import_module`, but I think for now we should keep it, especially since the Offline RL documentation uses that utility.

Actually, I think I can add this fix into the [Offline RL](https://docs.ray.io/en/latest/rllib/rllib-offline.html) docs PR, if that makes sense.

File: rllib/offline/estimators/qreg_torch_model.py
Comment (rapotdar) [inline rllib/offline/estimators/qreg_torch_model.py:149]: The earlier version of QReg was wrong here: k should start at t + 1, not t, and go all the way to t_prime (the subscripts in the [COBS](https://arxiv.org/pdf/1911.06854.pdf) paper are inclusive for the last index, not exclusive).

**Review by avnishn — COMMENTED** [review]
I think it'd be nice if we had a more intuitive way of specifying ope methods and their respective configs. I mentioned config objects, but I'm not opposed to other methods. 

I think its good that you've added integration tests, but something I'd hope to see are unit tests that force you to test some of the key crucial equations that you've written out. I think @sven1977 might have some different opinions here, and I'll defer to his judgement.

other than that, I think I have some nits here and there.

File: rllib/agents/trainer_config.py
Comment (rapotdar) [reply rllib/agents/trainer_config.py]: Yyes, that's the intended usage. Currently I kept support for {"is": {"type": "is"}} for backward compatibility, but I'll deprecate that in favour of only allowing either the full classpath or the actual class in the documentation PR.

File: rllib/offline/estimators/qreg_torch_model.py
Comment (rapotdar) [reply rllib/offline/estimators/qreg_torch_model.py:168]: Nope, I can switch it to numpy instead

File: doc/source/rllib/rllib-training.rst
Comment (gjoliver) [reply doc/source/rllib/rllib-training.rst:577]: wait, I don't get it, why do we double specify the name? "is" and ImportanceSampling?

File: rllib/algorithms/marwil/marwil.py
Comment (gjoliver) [reply rllib/algorithms/marwil/marwil.py:107]: ??
can you respond to @avnishn's comments please?

File: rllib/evaluation/rollout_worker.py
Comment (gjoliver) [reply rllib/evaluation/rollout_worker.py:355]: ??

File: rllib/offline/estimators/fqe_torch_model.py
Comment (gjoliver) [reply rllib/offline/estimators/fqe_torch_model.py:67]: ??

**Review by gjoliver — COMMENTED** [review]
does look very cool.
have 1 random question, plus I think some of Avnish's questions got dropped?

File: rllib/BUILD
Comment (rapotdar) [reply rllib/BUILD:1144]: Actually, looks like this string might be broken currently since we don't allow unknown configs, and the default is off_policy_estimation_methods = {}; I'll fix this before merge

File: rllib/algorithms/marwil/marwil.py
Comment (rapotdar) [reply rllib/algorithms/marwil/marwil.py:107]: So this is the pattern I was following, from Sven's comment earlier:
```
config:
  off_policy_estimators:
    my_is:
      type: ImportanceSampling
      ctor_kwarg1: [some value1]
    dr:
      type: DoublyRobust
      ctor_kwarg1: [some value]
      ctor_kwarg2: [some other value]
```

So the name e.g. `my_is` is the string under which the estimator's results are saved. `type` is the actual class or classpath for the estimator. For IS and WIS, this isn't really relevant, since they don't take any other arguments beyong policy and gamma.

However for e.g. DR, you can do something like this:
```
off_policy_estimation_methods: {
            "dr_qreg": {
                "type": DoublyRobust,
                "q_model_type": "qreg",
            },
            "dr_fqe_tau_1": {
                "type": DoublyRobust,
                "q_model_type": "fqe",
                "tau": 0.01,
            },
            "dr_fqe_tau_5": {
                "type": DoublyRobust,
                "q_model_type": "fqe",
                "tau": 0.05,
            },
}
```
And so on. I've actually updated the `test_ope` test accordingly, although it now takes ~45 minutes to run (Should come back down to ~10 minutes when we put in parallel eval/k_fold_cv/Ray AIR)

File: rllib/offline/estimators/fqe_torch_model.py
Comment (rapotdar) [reply rllib/offline/estimators/fqe_torch_model.py:67]: We have `assert policy.action_space == Discrete`, should we do an assertion on the model side as well?

File: rllib/algorithms/marwil/marwil.py
Comment (rapotdar) [reply rllib/algorithms/marwil/marwil.py:107]: @gjoliver ^ Does this answer your questions?

File: rllib/evaluation/rollout_worker.py
Comment (rapotdar) [reply rllib/evaluation/rollout_worker.py:355]: Hm, I'm not sure how we can put this into the TrainerConfig objects though, since it's inside the optional config dict for the OPE, plus we want to support arbitrary Q models and corresponding arguments. We can chat about this tomorrow maybe?

File: rllib/algorithms/marwil/marwil.py
Comment (gjoliver) [reply rllib/algorithms/marwil/marwil.py:107]: this makes a lot of sense. thanks.
I think we should document this though, for example, add your DR estimate configs as an example to the doc?
and here, I wonder if we can just say:
```
# Use importance sampling estimators for reward.
# off_policy_estimation_methods is a map from arbitrary string names to Off-Policy Evaluation configs.
# Metrics for each OPE config will show up under corresponding key name.
```
?
Thanks.

File: rllib/agents/trainer_config.py
Comment (ArturNiederfahrenhorst) [reply rllib/agents/trainer_config.py:945]: I think this is an actual issue, because tune checkpoints will not return said classes and they *must* return full module/class paths for this this model to be sustainable.

File: rllib/offline/estimators/fqe_torch_model.py
Comment (avnishn) [reply rllib/offline/estimators/fqe_torch_model.py:69]: no thats fine probably

**Review by sven1977 — APPROVED** [review]
Looks good to go now. Thanks for this great work @Rohan138 !

**Comment by sven1977** [comment]
Waiting for tests to pass ...

**Pull request merged**
